File: bindings/go/dstu/sign257_test.go
package dstu

import (
	"bytes"
	"testing"
)

// SigningKey257/VerifyingKey257 (crypto_sign257, DSTU 4145 m=257) - m=257 sibling of
// SigningKey/VerifyingKey (T-199/T-204). Correctness (round trip, determinism of the nonce
// derivation), rejection (wrong message/wrong key), misuse (invalid signing key -
// zero/out-of-range, wrong-length key/signature).

func TestSign257VerifyRoundTrips(t *testing.T) {
	signingKey, _ := GenerateSigningKey257()
	defer signingKey.Close()
	verifyingKey := signingKey.VerifyingKey()
	defer verifyingKey.Close()
	message := []byte("a message whose origin and integrity matter")
	sig := signingKey.Sign(message)
	ok, err := verifyingKey.Verify(message, sig)
	if err != nil {
		t.Fatal(err)
	}
	if !ok {
		t.Fatal("expected the signature to verify")
	}
}

func TestSign257IsDeterministic(t *testing.T) {
	signingKey, _ := GenerateSigningKey257()
	defer signingKey.Close()
	message := []byte("same message every time")
	a := signingKey.Sign(message)
	b := signingKey.Sign(message)
	if !bytes.Equal(a, b) {
		t.Fatal("expected the same signature twice")
	}
}

func TestSign257WrongMessageIsRejected(t *testing.T) {
	signingKey, _ := GenerateSigningKey257()
	defer signingKey.Close()
	verifyingKey := signingKey.VerifyingKey()
	defer verifyingKey.Close()
	sig := signingKey.Sign([]byte("original message"))
	ok, err := verifyingKey.Verify([]byte("a different message"), sig)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected verification to fail")
	}
}

func TestSign257WrongKeyIsRejected(t *testing.T) {
	signingKey, _ := GenerateSigningKey257()
	defer signingKey.Close()
	otherSigningKey, _ := GenerateSigningKey257()
	defer otherSigningKey.Close()
	otherVerifyingKey := otherSigningKey.VerifyingKey()
	defer otherVerifyingKey.Close()
	message := []byte("message")
	sig := signingKey.Sign(message)
	ok, err := otherVerifyingKey.Verify(message, sig)
	if err != nil {
		t.Fatal(err)
	}
	if ok {
		t.Fatal("expected verification to fail")
	}
}

func TestZeroSigningKey257IsRejected(t *testing.T) {
	if _, err := SigningKey257FromBytes(make([]byte, Sign257PrivateKeyBytes)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestWrongLengthSigningKey257IsRejected(t *testing.T) {
	if _, err := SigningKey257FromBytes(make([]byte, 5)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestWrongLengthVerifyingKey257IsRejected(t *testing.T) {
	if _, err := VerifyingKey257FromBytes(make([]byte, 5)); err == nil {
		t.Fatal("expected an error")
	}
}

func TestWrongLengthSignature257IsRejected(t *testing.T) {
	signingKey, _ := GenerateSigningKey257()
	defer signingKey.Close()
	verifyingKey := signingKey.VerifyingKey()
	defer verifyingKey.Close()
	if _, err := verifyingKey.Verify([]byte("message"), make([]byte, 5)); err == nil {
		t.Fatal("expected an error")
	}
}

File: bindings/php/tests/Sign257Test.php
<?php

declare(strict_types=1);

use PHPUnit\Framework\TestCase;

/**
 * crypto_sign257 (DSTU 4145 m=257) - m=257 sibling of crypto_sign (T-199/T-204). Correctness
 * (round trip, determinism of the nonce derivation), rejection (wrong message/wrong key), misuse
 * (invalid signing key - zero/out-of-range, wrong-length verifying key/signature).
 */
final class Sign257Test extends TestCase
{
    public function testRoundTripsSignVerify(): void
    {
        $signingKey = dstu_core_sign257_keygen();
        $verifyingKey = dstu_core_sign257_verifying_key($signingKey);
        $message = 'a message whose origin and integrity matter';
        $signature = dstu_core_sign257_message($signingKey, $message);
        $this->assertTrue(dstu_core_sign257_verify($verifyingKey, $message, $signature));
    }

    public function testIsDeterministic(): void
    {
        $signingKey = dstu_core_sign257_keygen();
        $message = 'same message every time';
        $this->assertSame(
            dstu_core_sign257_message($signingKey, $message),
            dstu_core_sign257_message($signingKey, $message)
        );
    }

    public function testRejectsWrongMessage(): void
    {
        $signingKey = dstu_core_sign257_keygen();
        $verifyingKey = dstu_core_sign257_verifying_key($signingKey);
        $signature = dstu_core_sign257_message($signingKey, 'original message');
        $this->assertFalse(dstu_core_sign257_verify($verifyingKey, 'a different message', $signature));
    }

    public function testRejectsWrongKey(): void
    {
        $signingKey = dstu_core_sign257_keygen();
        $otherVerifyingKey = dstu_core_sign257_verifying_key(dstu_core_sign257_keygen());
        $message = 'message';
        $signature = dstu_core_sign257_message($signingKey, $message);
        $this->assertFalse(dstu_core_sign257_verify($otherVerifyingKey, $message, $signature));
    }

    public function testRejectsAllZeroSigningKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_sign257_verifying_key(str_repeat("\x00", 33));
    }

    public function testRejectsWrongLengthSigningKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_sign257_message('too short', 'message');
    }

    public function testRejectsWrongLengthVerifyingKey(): void
    {
        $this->expectException(\ValueError::class);
        dstu_core_sign257_verify('too short', 'message', str_repeat("\x00", 66));
    }

    public function testRejectsWrongLengthSignature(): void
    {
        $signingKey = dstu_core_sign257_keygen();
        $verifyingKey = dstu_core_sign257_verifying_key($signingKey);
        $this->expectException(\ValueError::class);
        dstu_core_sign257_verify($verifyingKey, 'message', 'too short');
    }
}
